enhancements/update/cluster-version-operator-x509-upstream-trust.md

```diff
@@ -0,0 +1,202 @@
+	//    defined by the TLSConfig consumer.
+	// 3. Feeding the the resulting trust bundle to the appropriate
+	//    operand, as defined by the TLSConfig consumer.  This may involve
+	//    writing a ConfigMap in an operator-managed namespace that is then
+	//    volume-mounted into operand pods.
+	//
+	// The namespace for the ConfigMap referenced by trustedCA is
+	// "openshift-config". Here is an example ConfigMap (in yaml):
+	//
+	// apiVersion: v1
+	// kind: ConfigMap
+	// metadata:
+	//  name: user-ca-bundle
+	//  namespace: openshift-config
+	//  data:
+	//    ca-bundle.crt: |
+	//      -----BEGIN CERTIFICATE-----
+	//      Custom CA certificate bundle.
+	//      -----END CERTIFICATE-----
+	//
+	// +optional
+	TrustedCA *ConfigMapNameReference `json:"trustedCA,omitempty"`
+}
+```
+
+[`ClusterVersionSpec`][api-spec] will be extended to include the following new property:
+
+```go
+// upstreamTLS specifies settings for TLS connections to the upstream
+// update service.  If trustedCA is set, it is used directly and not
+// merged with the system default trust bundle.  If upstreamTLS is set,
+// even if none of its constituent properties are set, it takes
+// precedence over the proxy configuration.  To use the proxy
+// configuration for connections to the upstream update service, unset
+// upstreamTLS completely.
+// +optional
+UpstreamTLS *TLSConfig `json:"upstreamTLS,omitempty"
+```
+
+### User Stories
+
+#### Local update recommendation service with non-standard certificate authority
+
+Users running a local [update recommendation service][update-service] (such as [Cincinnati][cincinnati]) may use a serving certificate signed by an internal CA.
+This enhancement will allow the CVO to successfully pull recommended updates from those services.
+Connecting to the services over HTTP (without TLS) would be unwise, because an attacker who can man-in-the-middle the update recommendation service can suggest clusters do crazy things (e.g. "I see you're on 4.4.3.  I suggest you update to 4.1.0"), for which there are no CVO-side guards at the moment.
+And in some cases (e.g. updating to fumbled releases which we initially expected to be healthy), no CVO-side guards are possible.
+This enhancement enables trusted, secure connections to the update recommendation service without requiring users to provision a certificate signed by a default CA.
+
+#### Transparent, network-level proxies
+
+[rhbz#1773419][rhbz1773419] discusses issue when CVO upstream requests are not redirected via an `httpsProxy` setting, but are instead routed through a proxy via network-level configuration (e.g. all outgoing TCP traffic is routed through a proxy).
+The CVO currently only injects the Proxy config's `trustedCA` data [when `httpsProxy` is non-empty][trustedCA-vs-httpsProxy], so there is currently no mechanism for configuring the CVO's trusted CA store in this situation, short of forcing an explicit value into `httpsProxy`.
```

Comment:
the CVO ought to consume the proxy CAs as long as they are defined, regardless of whether httpsProxy is set or not.  That's a bug in the CVO's proxy config handling behavior.  (orthogonal to the use case being solved by this enhancement)

Reply:
> the CVO ought to consume the proxy CAs as long as they are defined, regardless of whether httpsProxy is set or not

I'm in favor of using the Proxy's `trustedCA` for transparent, network-level proxies.  But that has been [historically][1] [unpopular][2].  If you think we can build a consensus around consuming `trustedCA` regardless of `httpProxy` or `httpsProxy`, I'm happy to cover this case that way.

[1]: https://github.com/openshift/installer/pull/2658#issuecomment-554517800
[2]: https://github.com/openshift/installer/pull/3039

Reply:
that issue was closed because it was trying to use a field called "AdditionalCAs" in a way that wasn't well defined/aligned w/ the original intention.  I don't think it's closure really means "we should never use proxy CAs for network proxies", it just means "the installer needs to make it clear which CAs are for proxies and which CAs are for "additional use".

The concern w/ that PR is that it was taking a field that was previously only consumed for proxy CAs when a proxy config was defined, and now always consuming it, when the field doc did not make that usage clear.  It really had nothing to do with how the proxy CAs+config operated, just how the installer turned its inputs into a proxy CA/config.